kv: panic "use of tracing span after finish" in server tests related to replica rebalancer & multi-store

**Describe the problem**

I ran `dev test //pkg/server` and got the following error on one of the tests. The stack trace tells me it's something about the replica rebalancer (`kvserver.ResolveLBRebalancingObjective`).

```
panic: use of Span after Finish. Span: initialize-additional-stores. Finish previously called at: <stack not captured. Set debugUseAfterFinish> [recovered]
        panic: use of Span after Finish. Span: initialize-additional-stores. Finish previously called at: <stack not captured. Set debugUseAfterFinish>

goroutine 41541 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0x8acb42e90?, {0x1fa7360, 0x8bf7bd200})
        github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:229 +0x6a
panic({0x465720, 0x85c93dd80})
        GOROOT/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish(0x886d470?)
        github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:181 +0x12b
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).RecordingType(0x8acd2b400)
        github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:441 +0x1e
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).IsVerbose(...)
        github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:449
github.com/cockroachdb/cockroach/pkg/util/log.getSpanOrEventLog({0x1fa7360, 0x85d27c480})
        github.com/cockroachdb/cockroach/pkg/util/log/trace.go:92 +0x3a
github.com/cockroachdb/cockroach/pkg/util/log.logfDepthInternal({0x1fa7360, 0x85d27c480}, 0x2, 0x1, 0x0, 0x0?, {0xe110b3, 0x47}, {0x0, 0x0, ...})
        github.com/cockroachdb/cockroach/pkg/util/log/channels.go:101 +0x554
github.com/cockroachdb/cockroach/pkg/util/log.logfDepth(...)
        github.com/cockroachdb/cockroach/pkg/util/log/channels.go:39
github.com/cockroachdb/cockroach/pkg/util/log.Infof(...)
        github.com/cockroachdb/cockroach/bazel-out/freebsd-fastbuild/bin/pkg/util/log/log_channels_generated.go:277
github.com/cockroachdb/cockroach/pkg/kv/kvserver.ResolveLBRebalancingObjective({0x1fa7360, 0x85d27c480}, 0x85fd88000, 0x8ad542430?)
        github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/rebalance_objective.go:265 +0xce
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*RebalanceObjectiveManager).maybeUpdateRebalanceObjective(0x85f6bfbc0, {0x1fa7360, 0x85d27c480})
        github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/rebalance_objective.go:231 +0xdb
github.com/cockroachdb/cockroach/pkg/kv/kvserver.newRebalanceObjectiveManager.func3(_, {0x20000000, 0x20000000, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xbff0000000000000, ...}, ...)
        github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/rebalance_objective.go:211 +0x4c
github.com/cockroachdb/cockroach/pkg/kv/kvserver/allocator/storepool.(*StorePool).capacityChanged(_, _, {0x20000000, 0x20000000, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...)
        github.com/cockroachdb/cockroach/pkg/kv/kvserver/allocator/storepool/store_pool.go:1020 +0x177
github.com/cockroachdb/cockroach/pkg/kv/kvserver/allocator/storepool.(*StorePool).storeDescriptorUpdate(_, {0x7, {{0x0, 0x0, 0x0}}, {0x3, {{0x8ad5423e9, 0x3}, {0x8ad542410, 0xf}}, ...}, ...})
        github.com/cockroachdb/cockroach/pkg/kv/kvserver/allocator/storepool/store_pool.go:589 +0x4c7
github.com/cockroachdb/cockroach/pkg/kv/kvserver/allocator/storepool.(*StorePool).storeGossipUpdate(0x8ac52e0f0, {0x8ad5422e6?, 0x7?}, {{0x8bd092000, 0x14e, 0x160}, {0x1761b9ab3ac7e666, 0x0, 0x0}})
        github.com/cockroachdb/cockroach/pkg/kv/kvserver/allocator/storepool/store_pool.go:560 +0x1a5
github.com/cockroachdb/cockroach/pkg/gossip.(*infoStore).runCallbacks.func1()
        github.com/cockroachdb/cockroach/pkg/gossip/pkg/gossip/infostore.go:340 +0x74
github.com/cockroachdb/cockroach/pkg/gossip.newInfoStore.func1({0x8ae46b800?, 0x0?})
        github.com/cockroachdb/cockroach/pkg/gossip/pkg/gossip/infostore.go:188 +0x177
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
        github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:470 +0x146
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
        github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:461 +0x43b
```

**To Reproduce**

`dev test //pkg/server` on my machine. 

The other bits in the stack dump mention `server.TestNodesV2` so I assume the problem can be reproduced by stress-running that specific test.



Jira issue: CRDB-28193

--- a/pkg/kv/kvserver/rebalance_objective.go
+++ b/pkg/kv/kvserver/rebalance_objective.go
@@ -208,7 +208,9 @@ func newRebalanceObjectiveManager(
 	capacityChangeNotifier.SetOnCapacityChange(
 		func(storeID roachpb.StoreID, old, cur roachpb.StoreCapacity) {
 			if (old.CPUPerSecond < 0) != (cur.CPUPerSecond < 0) {
-				rom.maybeUpdateRebalanceObjective(ctx)
+				// NB: On capacity changes we don't have access to a context. Create a
+				// background context on callback.
+				rom.maybeUpdateRebalanceObjective(context.Background())
 			}
 		})
 
@@ -228,15 +230,16 @@ func (rom *RebalanceObjectiveManager) maybeUpdateRebalanceObjective(ctx context.
 	defer rom.mu.Unlock()
 
 	prev := rom.mu.obj
-	new := ResolveLBRebalancingObjective(ctx, rom.st, rom.storeDescProvider.GetStores())
+	next := ResolveLBRebalancingObjective(ctx, rom.st, rom.storeDescProvider.GetStores())
 	// Nothing to do when the objective hasn't changed.
-	if prev == new {
+	if prev == next {
 		return
 	}
 
-	log.Infof(ctx, "Updating the rebalance objective from %s to %s", prev.ToDimension(), new.ToDimension())
+	log.Infof(ctx, "Updating the rebalance objective from %s to %s",
+		prev.ToDimension(), next.ToDimension())
 
-	rom.mu.obj = new
+	rom.mu.obj = next
 	rom.mu.onChange(ctx, rom.mu.obj)
 }
 

--- a/pkg/kv/kvserver/rebalance_objective.go
+++ b/pkg/kv/kvserver/rebalance_objective.go
@@ -249,7 +249,7 @@ func (rom *RebalanceObjectiveManager) maybeUpdateRebalanceObjective(ctx context.
 	}
 
 	log.Infof(ctx, "Updating the rebalance objective from %s to %s",
-		prev.ToDimension(), next.ToDimension())
+		prev, next)
 
 	rom.mu.obj = next
 	rom.mu.onChange(ctx, rom.mu.obj)

--- a/pkg/kv/kvserver/rebalance_objective.go
+++ b/pkg/kv/kvserver/rebalance_objective.go
@@ -166,6 +166,7 @@ type gossipStoreCapacityChangeNotifier interface {
 // interface and registers a callback at creation time, that will be called on
 // a reblanace objective change.
 type RebalanceObjectiveManager struct {
+	log.AmbientContext
 	st                *cluster.Settings
 	storeDescProvider gossipStoreDescriptorProvider
 
@@ -181,20 +182,28 @@ type RebalanceObjectiveManager struct {
 
 func newRebalanceObjectiveManager(
 	ctx context.Context,
+	ambientCtx log.AmbientContext,
 	st *cluster.Settings,
 	onChange func(ctx context.Context, obj LBRebalancingObjective),
 	storeDescProvider gossipStoreDescriptorProvider,
 	capacityChangeNotifier gossipStoreCapacityChangeNotifier,
 ) *RebalanceObjectiveManager {
-	rom := &RebalanceObjectiveManager{st: st, storeDescProvider: storeDescProvider}
+	rom := &RebalanceObjectiveManager{
+		st:                st,
+		storeDescProvider: storeDescProvider,
+		AmbientContext:    ambientCtx,
+	}
+	rom.AddLogTag("rebalance-objective", nil)
+	ctx = rom.AnnotateCtx(ctx)
+
 	rom.mu.obj = ResolveLBRebalancingObjective(ctx, st, storeDescProvider.GetStores())
 	rom.mu.onChange = onChange
 
 	LoadBasedRebalancingObjective.SetOnChange(&rom.st.SV, func(ctx context.Context) {
-		rom.maybeUpdateRebalanceObjective(ctx)
+		rom.maybeUpdateRebalanceObjective(rom.AnnotateCtx(ctx))
 	})
 	rom.st.Version.SetOnChange(func(ctx context.Context, _ clusterversion.ClusterVersion) {
-		rom.maybeUpdateRebalanceObjective(ctx)
+		rom.maybeUpdateRebalanceObjective(rom.AnnotateCtx(ctx))
 	})
 	// Rather than caching each capacity locally, use the callback as a trigger
 	// to recalculate the objective. This is less expensive than recacluating
@@ -210,7 +219,9 @@ func newRebalanceObjectiveManager(
 			if (old.CPUPerSecond < 0) != (cur.CPUPerSecond < 0) {
 				// NB: On capacity changes we don't have access to a context. Create a
 				// background context on callback.
-				rom.maybeUpdateRebalanceObjective(context.Background())
+				cbCtx, span := rom.AnnotateCtxWithSpan(context.Background(), "capacity-change")
+				defer span.Finish()
+				rom.maybeUpdateRebalanceObjective(cbCtx)
 			}
 		})
 
@@ -229,6 +240,7 @@ func (rom *RebalanceObjectiveManager) maybeUpdateRebalanceObjective(ctx context.
 	rom.mu.Lock()
 	defer rom.mu.Unlock()
 
+	ctx = rom.AnnotateCtx(ctx)
 	prev := rom.mu.obj
 	next := ResolveLBRebalancingObjective(ctx, rom.st, rom.storeDescProvider.GetStores())
 	// Nothing to do when the objective hasn't changed.

--- a/pkg/kv/kvserver/rebalance_objective_test.go
+++ b/pkg/kv/kvserver/rebalance_objective_test.go
@@ -188,7 +188,9 @@ func TestRebalanceObjectiveManager(t *testing.T) {
 			callbacks = append(callbacks, obj)
 		}
 		return newRebalanceObjectiveManager(
-			ctx, st, cb, providerNotifier, providerNotifier,
+			ctx, log.MakeTestingAmbientCtxWithNewTracer(),
+			st, cb,
+			providerNotifier, providerNotifier,
 		), &callbacks
 	}
 

--- a/pkg/kv/kvserver/store.go
+++ b/pkg/kv/kvserver/store.go
@@ -1289,7 +1289,10 @@ func NewStore(
 		allocatorStorePool = cfg.StorePool
 		storePoolIsDeterministic = allocatorStorePool.IsDeterministic()
 
-		s.rebalanceObjManager = newRebalanceObjectiveManager(ctx, s.cfg.Settings,
+		s.rebalanceObjManager = newRebalanceObjectiveManager(
+			ctx,
+			s.cfg.AmbientCtx,
+			s.cfg.Settings,
 			func(ctx context.Context, obj LBRebalancingObjective) {
 				s.VisitReplicas(func(r *Replica) (wantMore bool) {
 					r.loadBasedSplitter.SetSplitObjective(

